tools/shared.py

```diff
@@ -2652,21 +2652,44 @@ def safe_copy(src, dst):
   if dst == '/dev/null': return
   shutil.copyfile(src, dst)
 
-def clang_preprocess(filename):
-  # TODO: REMOVE HACK AND PASS PREPROCESSOR FLAGS TO CLANG.
-  return subprocess.check_output([CLANG_CC, '-DFETCH_DEBUG=1', '-E', '-P', '-C', '-x', 'c', filename])
-
-def read_and_preprocess(filename):
-  f = open(filename, 'r').read()
-  pos = 0
-  include_pattern = re.compile('^#include\s*["<](.*)[">]\s?$', re.MULTILINE)
-  while(1):
-    m = include_pattern.search(f, pos)
-    if not m:
-      return f
-    included_file = open(os.path.join(os.path.dirname(filename), m.groups(0)[0]), 'r').read()
+def find_lines_with_preprocessor_directives(file):
+  return filter(lambda x: '#if' in x or '#elif' in x, open(file, 'r').readlines())
+
+def run_c_preprocessor_on_file(src, dst):
+  # Run LLVM's C preprocessor on the given file, expanding #includes and variables found in src/setting.js.
```

Comment:
I see what you're saying about the benefits of the LLVM preprocessor. It would be nice to use it, but I think we should only do so if (1) we can do it without hacks, and (2) we have one preprocessor for everything (or at least if we have 2 then they are simple and have identical behavior).

So I'm saying that if we want to use the LLVM preprocessor, let's replace the js one too. I'm not sure how that would work in terms of calling it from inside the js compiler though. Aside from that are the issues you mentioned with making it work without hacks, might need some refactoring of things (like no longer using `{{{ }}}` stuff, etc.). But maybe this could be done.

Alternatively, I still think a limited preprocessor - basically the one we have in js right now - is simple and gives 95% of the benefits. Instead of adding another preprocessor, we can run the existing one in js for our purposes here, we would just need a little refactoring to make it usable standalone so it is callable from python (this was proposed in #5296). This seems like very little work, but I admit it isn't as good a final result as using the LLVM preprocessor.